src/utils/Cache.ts

```diff
@@ -0,0 +1,246 @@
+import { AxiosRequestConfig, AxiosResponse } from 'axios';
+
+export type CacheTargets = CacheTarget[];
+export enum CacheTarget {
+  CACHE_API = 'CACHE_API',
+  MEMORY = 'MEMORY',
+}
+export type CacheResponse = {
+  data: any;
+  status: number;
+  statusText: string;
+  headers: any;
+};
+
+export const CACHE_API_KEY = 'sentinelhub-v1';
+export const SUPPORTED_TARGETS = [CacheTarget.CACHE_API, CacheTarget.MEMORY];
+const FALLBACK_TARGET = CacheTarget.MEMORY;
+
+export const memoryCache = new Map();
+
+// Factory will return an instance of a Cache interface.
+// The first availble target will be used
+// By default we will use cache_api if availble, if not we will cache in memory
+// user can also specify to just use in memory caching
+// If user provides a CacheTarget.CACHE_API and cache_api is not availble we will fallback to memory
+export function cacheFactory(optionalTargets: CacheTargets): ShCache {
+  const targets = optionalTargets || SUPPORTED_TARGETS;
+  const target = getFirstUsuableTarget(targets);
+  return constructCache(target);
+}
+
+function getFirstUsuableTarget(targets: CacheTargets): CacheTarget {
+  let firstTargetToUse = FALLBACK_TARGET; // default to memory if target is not supported
+  for (const key of targets) {
+    if (doesTargetExist(key)) {
+      firstTargetToUse = key;
+      break;
+    }
+  }
+  return firstTargetToUse;
+}
+
+export function doesTargetExist(target: CacheTarget): boolean {
+  switch (target) {
+    case CacheTarget.CACHE_API:
+      return Boolean(window.caches);
+    case CacheTarget.MEMORY:
+      return true;
+    default:
+      return false;
+  }
+}
+
+function constructCache(target: CacheTarget): ShCache {
+  switch (target) {
+    case CacheTarget.CACHE_API:
+      return new CacheApi();
+    case CacheTarget.MEMORY:
+      return new MemoryCache();
+    default:
+      throw new Error('Target not supported');
+  }
+}
+
+interface ShCache {
+  set(key: string, response: AxiosResponse): Promise<void>;
+  get(key: string, responseType: AxiosRequestConfig['responseType']): Promise<CacheResponse>;
+  has(key: string): Promise<boolean>;
+  keys(): Promise<string[]>;
+  delete(key: string): Promise<void>;
+  getHeaders(key: string): Promise<Record<string, any>>;
+  invalidate(): void;
+}
+
+class MemoryCache implements ShCache {
+  private cache: Record<string, any>;
+  public constructor() {
+    this.cache = memoryCache;
+  }
+
+  public async set(key: string, response: AxiosResponse): Promise<void> {
+    this.cache.set(key, response);
```

Comment:
Does this work? We are setting a `Response` here. When we retrieve it via `get`, we also read its `.data`. IIRC we can do this only once, so the next time we try to call `get` it would fail?

It might be a good idea to save an object instead of `Response`:
```
  this.cache.set(key, {
      data: cachedResponse.data,
      status: cachedResponse.status,
      statusText: cachedResponse.statusText,
      headers: cachedResponse.headers,
  });
```

A separate note: this scenario is a good candidate for testing ("what happens if I read the cached entry two times" - for memoy and for Cache API separately).

Reply:
Cache api requires a `Response` where as memory caching can be any javascript object, so here we are using the `AxiosResponse` directly as value, so in other words we do not have the issue of using a `Response` body twice when using the cache from memory

Reply:
I will add another test, just to cover our bases